### DISCUSSION TITLE
/articles/sharepoint-embedded-overview/

### DISCUSSION THREAD

**Discussion by giscus** [Voitanos/voitanos.io-comments#222] [Comments]
SharePoint Embedded, a new service from Microsoft 365, enables the creation of document centric apps with a custom UX but using SharePoint for document storage.
https://www.andrewconnell.com/articles/sharepoint-embedded-overview/

**Comment by palamundsen** [comment]
Hi Andrew. Thanks for an interesting article. In your example under "Customer facing document upload" you mention creating an application that targets end customers who need to upload files. Does this mean that my company can set up M365 with SP Embedded Storage and then invite external users to upload files to it? As long as I create an application that my external users have access to, then there's no need for any M365 subscription or license paid by the external users for accessing the data that sits in the Embedded Storage? In other words, the storage is configured and all costs are managed by my company?

**Reply by andrewconnell** [reply]
If I understand your question, that's correct. That's a single tenant scenario like what I mentioned in [that section](https://www.voitanos.io/blog/sharepoint-embedded-overview/#customer-facing-document-upload).

If *you* have the M365 subscription, and you create the web app that leverages SharePoint Embedded, then the *users* using your custom app won't pay for any storage, nor will they need any licenses to interact with the application that's storing documents in your M365 tenant. You are responsible for paying the metered usage of SharePoint Embedded (based on storage & API calls made by your app).

Does that answer your question?

**Reply by palamundsen** [reply]
Yes, it does, thanks! Embedded certainly looks like a feature that can cover scenarios that were difficult to address earlier

**Comment by Matej4386** [comment]
Hi Andrew. Really helpful article. Thank you very much. Just wondering:
1. what about files custom properties? How to define them? Would it be possible?
2. what about custom file access rights, for example many files in one container with different access rights. Does the new uploaded file inherits rights from container (like file in document library in SharePoint)?
3. In scenario that external users will have access to the container - what about security (in case of O365 users - login can be done through MSAL), but for external users the custom app should handle authentication. Maybe I just do not understand the concept.

**Reply by andrewconnell** [reply]
Great questions... let me try to answer but if something isn't clear, feel free to follow-up...

> 1. what about files custom properties? How to define them? Would it be possible?

You can define custom properties on a *Container*... that's done using the Microsoft Graph's `customProperties` endpoint on the container. You can learn more about that from the [beta docs here](https://learn.microsoft.com//graph/api/filestoragecontainer-post-customproperty?view=graph-rest-beta).

As for files, you can do anything you can normally do with a file in a document library because files in SPE Containers are no different than files in document libraries... they're just [driveItems](https://learn.microsoft.com/graph/api/resources/driveitem).

> 2. what about custom file access rights, for example many files in one container with different access rights. Does the new uploaded file inherits rights from container (like file in document library in SharePoint)?

Sort of... one difference between permissions in SharePoint vs. SPE is that you can't negate permissions. In other words, in SharePoint, say a user has WRITE permissions to a folder, but you've broken the inheritance of a file in that folder and they only have READER access on a file. You can't do that in SPE... you can only make it more permissive. For example, the user can be a READER on a folder, but can be a WRITER on a specific file. 

Think of it like this - you can ADD permissions at a more granular level, but you can't REMOVE/BLOCK permissions at a more granular level.

> 3. In scenario that external users will have access to the container - what about security (in case of O365 users - login can be done through MSAL), but for external users the custom app should handle authentication. Maybe I just do not understand the concept.

All users will login using the same Microsoft Entra tenant... external users can just be external/guests and don't need a M365/SP license to access the files. All access to the content is done through your custom app & all users must go through it. Depending on the setup of your app (single vs. multi-tenant), users can login from their tenant's Microsoft Entra account.

**Reply by Matej4386** [reply]
Thank you very much. All clear thanks to you. Have a great day.

**Comment by matthias-glubrecht** [comment]
Hi Andrew,
this sounds interesting, thanks for the article.
Given that the technical implementation of file storage seems to be similar to the storage of files in SharePoint document libraries, will it be possible to sync files from SharePoint embedded to a user's local drive using the OneDrive sync client?

**Reply by andrewconnell** [reply]
While I could be wrong, I don't think so. 

Access to the content is done through your Entra ID's application tied to your SPE container type. One of the points of SharePoint Embedded is for your app to control access to the content.
